### ISSUE TITLE
In factor vae: an inplace problem occurred when using loss.backward

### ISSUE THREAD

**Issue by scarletteshu** [AntixK/PyTorch-VAE#14]
the error message is:

`one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [4096, 6]], which is output 0 of TBackward, is at version 2; expected version 1 instead.`

it seems to be the problem that a term(`self.D_z_reserve`) used in D_tc_loss was calculated at vae_loss stage. 

`D_tc_loss = 0.5 * (F.cross_entropy(self.D_z_reserve, false_labels) +
                               F.cross_entropy(D_z_perm, true_labels))`

giving details:
I calculated and updated vae loss first （say `vae_loss.backward(retain_graph = True)`）and called its optimizer(say `vae_optim.step()`), then I want to update discriminator(say `D_tc_loss.backward()`).

and error message occurs as discribed at beginning.

when I delete term `F.cross_entropy(self.D_z_reserve, false_labels)` in D_tc_loss, everything goes alright.

**Comment by AntixK** [comment]
The FactorVAE as implemented here has this issue. I believe it is because of the way PyTorch Lightning calls the `loss_function` as the `self.D_Z_reserve` variable is updated by the model keeps track of the gradients for it before it was again updated. 

I think this can be easily rectified with the latent pytorch lightning version, where they have improved a lot of stuff. 

Using `detach()` removes any gradients that has been tracked. So when you call `backward()`, it does nothing to the loss that uses `self.D_z_reserve.detach()`. So, I wouldn't recommend using that.

**Comment by federicoromeo** [comment]
Still having this issue.
All other implementations have the same issue.

Has anyone solved it? Is it right to detach as suggested?

> The FactorVAE as implemented here has this issue. I believe it is because of the way PyTorch Lightning calls the `loss_function` as the `self.D_Z_reserve` variable is updated by the model keeps track of the gradients for it before it was again updated.
> 
> I think this can be easily rectified with the latent pytorch lightning version, where they have improved a lot of stuff.
> 
> Using `detach()` removes any gradients that has been tracked. So when you call `backward()`, it does nothing to the loss that uses `self.D_z_reserve.detach()`. So, I wouldn't recommend using that.
